File: 98-validate-binary-search-tree/validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    if(!root) return true;
    return dfsValid(root, -Infinity, Infinity);
};


function dfsValid(node, min, max){
    if(node.val <= min || node.val >= max){
        return false;
    }

    if(node.left){
        if(!dfsValid(node.left, min, node.val)){
            return false;
        }
    }

    if(node.right){
        if(!dfsValid(node.right, node.val, max)){
            return false;
        }
    }

    return true;
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.py

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, node: Optional[TreeNode]) -> bool:
        def isValid(node, min, max):
            if not node:
                return True
            if not (min < node.val < max):
                return False
            leftSub = isValid(node.left, min, node.val)
            rightSub = isValid(node.right, node.val, max)
            return leftSub and rightSub
        
        return isValid(node, float('-inf'), float('inf'))
